src/interfaces/chain.cpp

```diff
@@ -367,6 +367,13 @@ class ChainImpl : public Chain
     {
         return MakeUnique<RpcHandlerImpl>(command);
     }
+    void requestMempoolTransactions(Notifications& notifications) override
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/15652#issuecomment-477271519

> You might need to block waiting for previous notifications.

One way to do this:

```c++
void requestMempoolTransactions(Notifications& notifications) override
{
    std::promise<void> promise;
    CallFunctionInValidationInterfaceQueue([&] {
        LOCK2(::cs_main, ::mempool.cs);
        for (const CTxMemPoolEntry& entry : ::mempool.mapTx) {
            notifications.TransactionAddedToMempool(entry.GetSharedTx());
        }
        promise.set_value();
    });
    promise.get_future().wait();
}
```

Reply:
Previous suggestion in https://github.com/bitcoin/bitcoin/pull/15652/files#r269965497 isn't exception safe. Also, cs_main needs to be released before calling this.

Probably better:

```c++
void requestMempoolTransactions(Notifications& notifications) override
{
    std::packaged_task<void()> task([&] {
        LOCK2(::cs_main, ::mempool.cs);
        for (const CTxMemPoolEntry& entry : ::mempool.mapTx) {
            notifications.TransactionAddedToMempool(entry.GetSharedTx());
        }
    });
    CallFunctionInValidationInterfaceQueue([&] { task(); });
    task.get_future().get();
}
```

Reply:
Thanks @ryanofsky!

Reply:
Haven't finished but it looks like there will be a lock order issue - cs_main -> mempool.cs -> cs_main -> cs_wallet.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15652#discussion_r270050754

> Haven't finished but it looks like there will be a lock order issue - cs_main -> mempool.cs -> cs_main -> cs_wallet.

That does seem like a problem. One way to fix it would be to release main and mempool locks before calling TransactionAddedToMempool, by pushing the transactions to a vector, releasing the locks and then looping over the vector.

Alternately if you changed the function to `pushMempoolTransactions(std::function<void(std::vector<CTransactionRef> push_fn))` you could release the locks and then send the vector to the wallet directly. Both of these changes would be nice optimizations for #10102, since passing a vector once across the socket is more efficient than calling `TransactionAddedToMempool` repeatedly and requiring lots of roundtrip traffic.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15652#discussion_r269965497

Current code looks ok, but I think I was misguided in making this thread because previous code in d94847f20c750d88a0aefa572155d54f8e39c144 was also ok. I think:

1) I was right to point out there's a race condition where normal mempool add/remove notifications triggered before the `requestMempoolTransactions` call could arrive out of order during and after the call
2) I was wrong to assume that the race condition would actually cause a problem. There is no problem if an unknown transaction is removed or an already added transaction is added again.
3) I was wrong to think that my proposed changes would actually prevent add/remove notifications from arriving out of order. In fact, they would only prevent older add/remove notifications from arriving _during_ the `requestMempoolTransactions` call, not _after_ the call.

So, I think if you want go back to the simpler implementation in d94847f20c750d88a0aefa572155d54f8e39c144 that would be fine.

Or if you did want to prevent obsolete add/remove notifications from arriving _after_ the `requestMempoolTransactions` call in addition to _during_, you'd need acquire the locks before calling `CallFunctionInValidationInterfaceQueue` and also fill the vector at the same time:

```c++
void requestMempoolTransactions(Notifications& notifications) override
{
    std::vector<CTransactionRef> transactions;
    std::packaged_task<void()> task([&] {
        for (const CTransactionRef& tx : transactions) {
            notifications.TransactionAddedToMempool(tx);
        }
    });
    {
        LOCK2(::cs_main, ::mempool.cs);
        for (const CTxMemPoolEntry& entry : ::mempool.mapTx) {
            transactions.push_back(entry.GetSharedTx());
        }
        CallFunctionInValidationInterfaceQueue([&] { task(); });
    }
    AssertLockNotHeld(::cs_main);
    AssertLockNotHeld(::mempool.cs);
    task.get_future().get();
}
```

Reply:
> I was wrong to assume that the race condition would actually cause a problem. There is no problem if an unknown transaction is removed or an already added transaction is added again.

Right, just unnecessary extra work, which is not an issue I think.